www/updateUes.php

```diff
@@ -36,91 +57,217 @@
+                )) {
+                    $uesFound = true;
+                    break;
+                }
+            }
+
+            if ($uesFound) {
+                break;
+            }
+        } elseif (is_string($attrValue)) {
+            if (doUpdate(
+                $adapter,
+                $extSourceName,
+                $attrValue,
+                $attrMap,
+                $attrsToConversion,
+                $perunUserId,
+                $attributesFromIdP
+            )) {
+                $uesFound = true;
+                break;
+            }
+        }
+    }
+
+    if (!$uesFound) {
         throw new Exception(
-            'perun/www/updateUes.php: there is no UserExtSource with ExtSource ' . $attributesFromIdP['sourceIdPEntityID'][0] . ' and Login ' . $attributesFromIdP['sourceIdPEppn'][0]
+            CLASS_PREFIX . 'There is no UserExtSource that could be used for user ' . $perunUserId . ' and extSourcewith ExtSource ' . $attributesFromIdP[SOURCE_IDP_ENTITY_ID][0]
         );
     }
+} catch (\Exception $ex) {
+    Logger::warning(
+        CLASS_PREFIX . 'Updating UES for user with userId: ' . $perunUserId . ' was not successful: ' .
+        $ex->getMessage()
+    );
+}
+
+/**
+ * @param mixed $adapter
+ * @param mixed $extSourceName
+ * @param mixed $extLogin
+ * @param mixed $attrMap
+ * @param mixed $attrsToConversion
+ * @param mixed $perunUserId
+ * @param mixed $attributesFromIdP
+ */
+function doUpdate(
+    $adapter,
+    $extSourceName,
+    $extLogin,
+    $attrMap,
+    $attrsToConversion,
+    $perunUserId,
+    $attributesFromIdP
+): bool {
+    $userExtSource = getUserExtSource($adapter, $extSourceName, $extLogin);
+    if (null === $userExtSource) {
+        return false;
+    }
+
+    $attrsFromPerun = getAttributesFromPerun($adapter, $attrMap, $userExtSource);
+    if (empty($attrsFromPerun)) {
+        return false;
+    }
+
+    $attrsToUpdate = getAttributesToUpdate($attrsFromPerun, $attrsToConversion, $attributesFromIdP);
+    if (empty($attrsToUpdate)) {
+        return false;
+    }
+
+    return updateUserExtSource($adapter, $attrsToUpdate, $userExtSource, $perunUserId);
+}
+
+function getUserExtSource($adapter, $extSourceName, $extLogin)
+{
+    try {
+        return $adapter->getUserExtSource($extSourceName, $extLogin);
+    } catch (SimpleSAML\Module\perun\Exception $ex) {
+        Logger::debug(CLASS_PREFIX . 'Caught exception when fetching user ext source, probably does not exist.');
+        Logger::debug(CLASS_PREFIX . $ex->getMessage());
+
+        return null;
+    }
+}
 
-    $attributesFromPerunRaw = $adapter->getUserExtSourceAttributes($userExtSource['id'], array_keys($attrMap));
+function getAttributesFromPerun($adapter, $attrMap, $userExtSource): array
+{
+    $attributesFromPerunRaw = $adapter->getUserExtSourceAttributes($userExtSource[ID], array_keys($attrMap));
     $attributesFromPerun = [];
+
     foreach ($attributesFromPerunRaw as $attributeFromPerunRaw) {
```

Comment:
Please change the name of the variable containing items to something different, (i.e. `$rawAttribute`). It is really hard not to overlook the difference in the naming. I believe that with the IDE suggestions someone might pretty quick use the container variable instead of the item.